### PR TITLE
More typing

### PR DESCRIPTION
**Pull request by bryevdv**
This PR adds type annotations to 
* `legate.core`
* `legate.core.communicator`
* `legate.core.io`

After this PR all that remain is `runtime.py`, `solver.py` and `store.py`

### CODE REVIEW

File: legate/core/io.py [hunk 1]
```diff
@@ -185,41 +199,42 @@ def ingest(
...
+        legion.legion_sharding_functor_invert(
+            sid,
+            shard,
+            domain,
+            domain,
+            total_shards,
+            points_ptr,
+            points_size,
+        )
+        points = []
```
Comment (magnatelee) [inline legate/core/io.py:229]: maybe it's outside the scope of this PR, but I'd prefer a for-comprehension here than an explicit loop calling `append`.

File: legate/core/legate.py [hunk 2]
```diff
@@ -80,7 +80,7 @@ def stores(self) -> list[Store]:
     @staticmethod
     def from_stores(
         dtype: pyarrow.DataType,
-        stores: list[Store],
+        stores: list[Optional[Store]],
```
Comment (magnatelee) [inline legate/core/legate.py:83]: It's actually a bit surprising that you figured this out just by making mypy happy. was there any call site that required this type (which btw is correct)?

File: legate/core/store.py [hunk 3]
```diff
@@ -1197,7 +1197,7 @@ def compute_key_partition(self, restrictions):
                 partition = Tiling(self._runtime, tile_shape, launch_shape)
             return partition
 
-    def compute_projection(self, proj_fn=None, launch_ndim=None):
+    def compute_projection(self, proj_fn=None, launch_ndim=None) -> Any:
```
Comment (magnatelee) [inline legate/core/store.py:1200]: I'm pretty sure this returns integers.

File: legate/core/store.py [hunk 4]
```diff
@@ -1225,7 +1225,9 @@ def compute_projection(self, proj_fn=None, launch_ndim=None):
     def find_restrictions(self):
         return self._transform.convert_restrictions(self._storage.restrictions)
 
-    def find_or_create_legion_partition(self, partition, complete=False):
+    def find_or_create_legion_partition(
+        self, partition, complete=False
+    ) -> Any:
```
Comment (magnatelee) [inline legate/core/store.py:1230]: if we could use `legion.legion_logical_partition_t` here, that's the right type.

File: legate/core/legate.py (hunk 2 above)
Comment (bryevdv) [reply legate/core/legate.py:83]: Line 143 of `io.py` does this
```
  array = Array(dtype, [None, store])
```

File: legate/core/store.py (hunk 4 above)
Comment (bryevdv) [reply legate/core/store.py:1230]: Not presently. That will be better to sort out once any cffi vs cython dust settles.

File: legate/core/io.py (hunk 1 above)
Comment (bryevdv) [reply legate/core/io.py]: I'm generally a fan of going small-scale code gardening as it is noticed https://github.com/nv-legate/legate.core/pull/196/commits/83b6367b1feadcd79dc2d163a3dc22e8dcd82322

File: legate/core/store.py (hunk 3 above)
Comment (bryevdv) [reply legate/core/store.py]: https://github.com/nv-legate/legate.core/pull/196/commits/6a5897c32657ab74b13d989d99e01b2a577a5107

**Review by magnatelee — APPROVED** [review]

**Pull request merged**